Add relations recipe (#391)
kysely-org/kysely: site/docs/recipes/relations.md

# Relations

> Kysely is just a query builder and doesn't have the concept of relations. Kysely builds the SQL you tell it to and nothing more. This doesn't mean there's no way to nest related rows in your queries. You just have to do it with the tools SQL and the dialect provide. In this recipe we show one way to do that on PostgreSQL.

Reviewer: ```suggestion
<h3>
Kysely IS NOT an ORM. Kysely DOES NOT have the concept of relations.
Kysely IS a query builder. Kysely DOES build the SQL you tell it to, nothing more, nothing less.
</h3>

Phew, glad we got that out the way..

All that was said above doesn't mean there's no way to nest related rows in your queries. 
You just have to do it with the tools SQL and the underlying dialect (e.g. PostgreSQL) you're using provide. 
In this recipe we show one way to do that when using the PostgreSQL built-in dialect.
```

## PostgreSQL `jsonb` data type and functions

> PostgreSQL has a rich JSON support through its `json` and `jsonb` data types and functions. The node PostgreSQL driver `pg` automatically parses returned `json` and `jsonb` columns as objects. With the combination of these two things, we can write some super efficient queries with nested relations.

Reviewer: ```suggestion
PostgreSQL has rich JSON support through its `json` and `jsonb` data types and functions. `pg`, the node PostgreSQL driver, automatically parses returned `json` and `jsonb` columns as json objects. With the combination of these two things, we can write some super efficient queries with nested relations.
```

> Let's start with the raw SQL and then see how we can write the query using Kysely in a nice type-safe way.

Reviewer: ```suggestion
Let's start with some raw SQL, and then see how we can write the query using Kysely in a nice type-safe way.
```

> In the following query, we fetch a list of persons and for each person, we nest the person's pets and the person's mother into the returned objects:

Reviewer: ```suggestion
In the following query, we fetch a list of people (from "person" table) and for each person, we nest the person's pets, and mother, into the returned objects:
```

```ts
SELECT
  person.*,

  // Select person's pets as a jsonb array
  (
    SELECT 
      COALESCE(JSONB_AGG(pets), '[]')
    FROM
    (
      SELECT 
        pet.id, pet.name
      FROM
        pet
      WHERE 
        pet.owner_id = person.id
      ORDER BY 
        pet.name
    ) pets
  ) pets,

  // Select person's mother as a jsonb object
  (
    SELECT 
      TO_JSONB(mother)
    FROM
    (
      SELECT 
        mother.*
      FROM
        person as mother
      WHERE 
        mother.id = person.mother_id
    ) mother
  ) mother
FROM
  person
```

> Simple right 😅. Yeah, not so much. But it does provide 100% control over the queries and a really good performance as long as you have indices for `pet.owner_id` and `person.mother_id`.

Reviewer: ```suggestion
Simple right 😅. Yeah, not so much. But it does provide 100% control over the queries and a really good performance as long as you have indices (or indexes, we don't judge) for "pet.owner_id" and "person.mother_id".
```

> Fortunately we can improve and simplify this a lot using kysely. First let's define couple of helpers:

Reviewer: ```suggestion
Fortunately we can improve and simplify this a lot using Kysely. First let's define a couple of helpers:
```

```ts
function jsonArrayFrom<O>(
  expr: Expression<O>
): RawBuilder<Simplify<O>[]> {
  return sql`(select coalesce(jsonb_agg(agg), '[]') from ${expr} as agg)`
}

export function jsonObjectFrom<O>(
  expr: Expression<O>
): RawBuilder<Simplify<O>> {
  return sql`(select to_jsonb(obj) from ${expr} as obj)`
}
```

> These helpers are included in kysely and you can import them from the `helpers` package like this:

Reviewer: ```suggestion
These helpers are included in Kysely and you can import them from the `helpers` module like this:
```

Author: Too much golang 😅

# Relations

<h3>
Kysely IS NOT an ORM. Kysely DOES NOT have the concept of relations.
Kysely IS a query builder. Kysely DOES build the SQL you tell it to, nothing more, nothing less.
</h3>

Phew, glad we got that out the way..

All that was said above doesn't mean there's no way to nest related rows in your queries. 
You just have to do it with the tools SQL and the underlying dialect (e.g. PostgreSQL) you're using provide. 
In this recipe we show one way to do that when using the PostgreSQL built-in dialect.

## PostgreSQL `jsonb` data type and functions

PostgreSQL has rich JSON support through its `json` and `jsonb` data types and functions. `pg`, the node PostgreSQL driver, automatically parses returned `json` and `jsonb` columns as json objects. With the combination of these two things, we can write some super efficient queries with nested relations.

Let's start with some raw SQL, and then see how we can write the query using Kysely in a nice type-safe way.

> In the following query, we fetch a list of people (from "person" table) and for each person, we nest the person's pets, and mother, into the returned objects:

Reviewer 2: ```suggestion
In the following query, we fetch a list of people (from "person" table) and for each person, we nest the person's pets and owner, into the returned objects:
```

> ```ts

Reviewer 2: ```suggestion
```sql
```

Author: Oof, nice catch!

SELECT
  person.*,

  // Select person's pets as a jsonb array
  (
    SELECT 
      COALESCE(JSONB_AGG(pets), '[]')
    FROM
    (
      SELECT 
        pet.id, pet.name
      FROM
        pet
      WHERE 
        pet.owner_id = person.id
      ORDER BY 
        pet.name
    ) pets
  ) pets,

>   // Select person's mother as a jsonb object

Reviewer 3: ```suggestion
  # Select person's mother as a jsonb object
```

Author: Good catch, actually should be `--`
